runtimes/shared-configuration/src/fee.rs

```diff
@@ -148,3 +161,104 @@ where
+	FeeCreditor: HandleCredit<Runtime::AccountId, Runtime::Fungibles>,
+	TipCreditor: HandleCredit<Runtime::AccountId, Runtime::Fungibles>,
+	AssetIdOf<Runtime>: FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default + Eq + TypeInfo,
+{
+	type AssetId = AssetIdOf<Runtime>;
+	type Balance = BalanceOf<Runtime>;
+	type LiquidityInfo = fungibles::Credit<Runtime::AccountId, Runtime::Fungibles>;
+
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &Runtime::AccountId,
+		_call: &Runtime::RuntimeCall,
+		_info: &DispatchInfoOf<Runtime::RuntimeCall>,
+		asset_id: Self::AssetId,
+		fee: Self::Balance,
+		_tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		// We don't know the precision of the underlying asset. Because the converted fee could be
+		// less than one (e.g. 0.5) but gets rounded down by integer division we introduce a minimum
+		// fee.
+		let min_converted_fee = if fee.is_zero() { Zero::zero() } else { One::one() };
+		let converted_fee = Converter::to_asset_balance(fee, asset_id)
+			.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?
+			.max(min_converted_fee);
+		let can_withdraw =
+			<Runtime::Fungibles as Inspect<Runtime::AccountId>>::can_withdraw(asset_id, who, converted_fee);
+		if can_withdraw != WithdrawConsequence::Success {
+			return Err(InvalidTransaction::Payment.into())
+		}
+		<Runtime::Fungibles as fungibles::Balanced<Runtime::AccountId>>::withdraw(
+			asset_id,
+			who,
+			converted_fee,
+			Exact,
+			Protect,
+			Polite,
+		)
+		.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))
+	}
+
+	/// Note: The `corrected_fee` already includes the `tip`.
+	fn correct_and_deposit_fee(
+		who: &Runtime::AccountId,
+		_dispatch_info: &DispatchInfoOf<Runtime::RuntimeCall>,
+		_post_info: &PostDispatchInfoOf<Runtime::RuntimeCall>,
+		corrected_fee: Self::Balance,
+		tip: Self::Balance,
+		paid: Self::LiquidityInfo,
+	) -> Result<(AssetBalanceOf<Runtime>, AssetBalanceOf<Runtime>), TransactionValidityError> {
+		let min_converted_fee = if corrected_fee.is_zero() { Zero::zero() } else { One::one() };
+		// Convert the corrected fee and tip into the asset used for payment.
+		let converted_fee = Converter::to_asset_balance(corrected_fee, paid.asset())
+			.map_err(|_| -> TransactionValidityError { InvalidTransaction::Payment.into() })?
+			.max(min_converted_fee);
+		let converted_tip = Converter::to_asset_balance(tip, paid.asset())
+			.map_err(|_| -> TransactionValidityError { InvalidTransaction::Payment.into() })?;
+
+		// Calculate how much refund we should return.
+		let (final_fee, refund) = paid.split(converted_fee);
+		// Split the tip from the fee
+		let (final_fee_minus_tip, final_tip) = final_fee.split(converted_tip);
+
+		let _ = <Runtime::Fungibles as fungibles::Balanced<Runtime::AccountId>>::resolve(who, refund);
+
+		FeeCreditor::handle_credit(final_fee_minus_tip);
+		TipCreditor::handle_credit(final_tip);
```

Comment:
This should be the main difference with the Parity's implementation, and it looks good, let's be sure to test the tip handling in the integration tests